### PR TITLE
Marshal and un-marshal objects while save/read to/from etcd

### PR DESCRIPTION
**Pull request by shtripat**
tendrl-bug-id: Tendrl/commons#893
Signed-off-by: Shubhendu <shtripat@redhat.com>

### CODE REVIEW

**Comment by shtripat** [comment]
THIS IS WIP PR. PLEASE DONT MERGE.

File: tendrl/commons/objects/definition/master.yaml [hunk 1]
```diff
@@ -1110,6 +1112,9 @@ namespace.tendrl:
         timeout:
           help: Job is timed out after 10 mins if yes
           type: String
+        valid_util:
```
[the remark this answers is not in the document]
Comment (shtripat) [reply tendrl/commons/objects/definition/master.yaml]: Ack. will change this

[the remark this answers is not in the document]
Comment (shtripat) [reply tendrl/commons/objects/definition/master.yaml]: `timeout` is a flag which says if timeout functionality is allowed for entity. `valid_until` is used to decide the validity period for the entity

File: tendrl/commons/jobs/__init__.py [hunk 2]
```diff
@@ -169,17 +164,14 @@ def process_job(job):
...
             return
 
-        job_status_key = "/queue/%s/status" % job.job_id
-        job_lock_key = "/queue/%s/locked_by" % job.job_id
         try:
             lock_info = dict(node_id=NS.node_context.node_id,
                              fqdn=NS.node_context.fqdn,
                              tags=NS.node_context.tags,
                              type=NS.type)
-            etcd_utils.write(job_status_key, "processing",
```
Comment (r0h4n) [inline tendrl/commons/jobs/__init__.py]: Please use these etcd_util writes here, we need that `prevValue="new"` check ,

make the `status` and `locked_by` watchable

File: tendrl/commons/jobs/__init__.py [hunk 3]
```diff
@@ -217,9 +209,8 @@ def process_job(job):
             )
             the_flow.run()
             try:
-                etcd_utils.write(job_status_key,
-                                 "finished",
-                                 prevValue="processing")
```
Comment (r0h4n) [inline tendrl/commons/jobs/__init__.py]: Please use these etcd_util writes here, we need that `prevValue="new"` check ,

make the `status` and `locked_by` watchable

File: tendrl/commons/objects/__init__.py [hunk 4]
```diff
@@ -189,71 +152,35 @@ def load(self):
...
-                    else:
-                        setattr(_copy, item['name'], dct)
+        self.render()
+        key = self.value + '/data'
+        try:
+            val_str = NS._int.client.read(key).value
+        except etcd.EtcdKeyNotFound:
+            return self
+        loc_dict = json.loads(val_str)
+        for attr_name, attr_val in vars(self).iteritems():
```
[the remark this answers is not in the document]
Comment (shtripat) [reply tendrl/commons/objects/__init__.py]: Ok. will add that.

File: tendrl/commons/objects/node/__init__.py [hunk 5]
```diff
@@ -2,14 +2,17 @@
 
 
 class Node(objects.BaseObject):
-    def __init__(self, fqdn=None,
+    def __init__(self, node_id=None, fqdn=None,
```
[the remark this answers is not in the document]
Comment (shtripat) [reply tendrl/commons/objects/node/__init__.py]: Ok. Should we take up this in next PR as its quite huge now?

File: tendrl/commons/jobs/__init__.py [hunk 6]
```diff
@@ -109,9 +105,8 @@ def process_job(job):
                 # mark status as "failed" and Job.error =
                 # "Timed out"
                 try:
-                    etcd_utils.write(job_status_key,
```
[the remark this answers is not in the document]
Comment (shtripat) [reply tendrl/commons/jobs/__init__.py]: I can make status and locked_by watchable attributes I feel. Will try that.

File: tendrl/commons/jobs/__init__.py (hunk 2 above)
Comment (shtripat) [reply tendrl/commons/jobs/__init__.py]: I can make status and locked_by watchable attributes I feel. Will try that.

File: tendrl/commons/jobs/__init__.py (hunk 3 above)
Comment (shtripat) [reply tendrl/commons/jobs/__init__.py]: I can make status and locked_by watchable attributes I feel. Will try that.

File: tendrl/commons/objects/__init__.py [hunk 7]
```diff
@@ -83,102 +82,66 @@ def load_definition(self):
...
-        self.render()
         if "Message" not in self.__class__.__name__:
             # If local object.hash is equal to
             # central_store object.hash, return
             if self.hash_compare_with_central_store(ttl=ttl):
                 return
-        if update:
-            current_obj = self.load()
-            for attr, val in vars(self).iteritems():
-                if isinstance(val, (types.FunctionType,
```
[the remark this answers is not in the document]
Comment (shtripat) [reply tendrl/commons/objects/__init__.py]: After putting back this check few of objects fail while save.I would suggest to do these changes later and everything else would be broken due to this.

File: tendrl/commons/objects/__init__.py [hunk 8]
```diff
@@ -83,102 +82,66 @@ def load_definition(self):
...
-            # convert list, dict (json) to python based on definitions
-            _type = self._defs.get("attrs", {}).get(item['name'],
-                                                    {}).get("type")
-            if _type:
-                if _type.lower() in ['json', 'list']:
-                    if item['value']:
+        watchables = self._defs.get("watch_attrs", [])
+        if self.__class__.__name__ in ['Config', 'Definition'] or \
+            len(watchables) > 0:
+            for item in self.render():
```
[the remark this answers is not in the document]
Comment (r0h4n) [reply tendrl/commons/objects/__init__.py]: rendered_obj = self.render()  ??

for item in rendered_obj

**Comment by r0h4n** [comment]
Good job @shtripat , 

please fix pep8 

and unit tests when possible

**Pull request merged**